url/src/main/java/com/datastax/oss/dsbulk/url/S3URLStreamHandler.java

```diff
@@ -62,9 +80,67 @@ public InputStream getInputStream() {
       LOGGER.debug("Getting S3 input stream for object '{}' in bucket '{}'...", key, bucket);
       GetObjectRequest getObjectRequest =
           GetObjectRequest.builder().bucket(bucket).key(key).build();
+      String query = url.getQuery();
+      if (query == null) {
+        throw new IllegalArgumentException(
+            "You must provide S3 client credentials in the URL query parameters.");
+      }
+      S3Client s3Client = s3ClientCache.computeIfAbsent(query, this::getS3Client);
+      return getInputStream(s3Client, getObjectRequest);
+    }
+
+    @VisibleForTesting
+    InputStream getInputStream(S3Client s3Client, GetObjectRequest getObjectRequest) {
       return s3Client.getObjectAsBytes(getObjectRequest).asInputStream();
     }
 
+    @VisibleForTesting
+    S3Client getS3Client(String query) {
+      Map<String, String> parameters = parseParameters(query);
+      String region;
+      if (parameters.containsKey(REGION)) {
+        region = parameters.get(REGION);
+      } else {
+        throw new IllegalArgumentException("You must supply an AWS 'region' parameter on S3 URls.");
+      }
+
+      SdkHttpClient httpClient = UrlConnectionHttpClient.builder().build();
+      S3ClientBuilder builder = S3Client.builder().httpClient(httpClient).region(Region.of(region));
+
+      String profile = parameters.getOrDefault(PROFILE, null);
+      String accessKeyId = parameters.getOrDefault(ACCESS_KEY_ID, null);
+      String secretAccessKey = parameters.getOrDefault(SECRET_ACCESS_KEY, null);
+      if (!StringUtils.isBlank(profile)) {
+        LOGGER.info("Using AWS profile {} to connect to S3.", profile);
+        builder.credentialsProvider(ProfileCredentialsProvider.create(profile));
+      } else if (accessKeyId != null || secretAccessKey != null) {
+        LOGGER.warn("Using access key ID and secret access key to connect to S3.");
+        LOGGER.warn(
+            "This is considered an insecure way to pass credentials. Please use a 'profile' instead.");
+        if (accessKeyId == null || secretAccessKey == null) {
+          throw new IllegalArgumentException(
+              "Both 'accessKeyId' and 'secretAccessKey' must be present if either one is provided.");
+        }
+        builder.credentialsProvider(() -> AwsBasicCredentials.create(accessKeyId, secretAccessKey));
+      } else {
+        LOGGER.info("Using default credentials provider to connect to S3.");
+      }
+
+      return builder.build();
+    }
+
+    private Map<String, String> parseParameters(String query) {
+      Map<String, String> parameters = new HashMap<>();
+      String[] paramList = query.split("&", 0);
+      for (String param : paramList) {
+        String[] parts = param.split("=", 0);
+        if (parts.length > 1 && !StringUtils.isBlank(parts[1])) {
+          parameters.put(parts[0], parts[1]);
```

Comment:
You need to take care of decoding the parameter keys and values, see https://stackoverflow.com/questions/13592236/parse-a-uri-string-into-name-value-collection for an example of how to do it.

Reply:
I did see that SO post, but it seemed like overkill for parsing a string that comes from "responsible" users expected to provide valid input. Is that assumption too unsafe? I would prefer an existing library (like the `URLEncodedUtils` from answer 2) to writing anything more complicated that what's here, but I'll defer to you if you'd like something different. :)

Reply:
adding httpclient as a dependency makes me uncomfortable, mostly because this library has been a frequent target for CVEs.